core/grant/service.go

```diff
@@ -257,3 +270,176 @@ func (s *Service) expiredInActiveUserAccess(ctx context.Context, timeLimiter cha
+		return nil, fmt.Errorf("fetching access from provider: %w", err)
+	}
+
+	resourceConfigMap := make(map[string]*domain.ResourceConfig)
+	for _, rc := range p.Config.Resources {
+		resourceConfigMap[rc.Type] = rc
+	}
+
+	resourcesMap := make(map[string]*domain.Resource)
+	for _, r := range resources {
+		resourcesMap[r.URN] = r
+	}
+
+	activeGrants, err := s.repo.List(ctx, domain.ListGrantsFilter{
+		ProviderTypes: []string{p.Type},
+		ProviderURNs:  []string{p.URN},
+		ResourceTypes: criteria.ResouceTypes,
+		ResourceURNs:  criteria.ResourceURNs,
+		Statuses:      []string{string(domain.GrantStatusActive)},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting active grants: %w", err)
+	}
+	// map[resourceURN]map[accounttype:accountId]map[permissionsKey]grant
+	activeGrantsMap := map[string]map[string]map[string]*domain.Grant{}
+	for _, g := range activeGrants {
+		if activeGrantsMap[g.Resource.URN] == nil {
+			activeGrantsMap[g.Resource.URN] = map[string]map[string]*domain.Grant{}
+		}
+
+		accountSignature := getAccountSignature(g.AccountType, g.AccountID)
+		if activeGrantsMap[g.Resource.URN][accountSignature] == nil {
+			activeGrantsMap[g.Resource.URN][accountSignature] = map[string]*domain.Grant{}
+		}
+
+		activeGrantsMap[g.Resource.URN][accountSignature][g.PermissionsKey()] = &g
+	}
+
+	var newGrants []*domain.Grant
+	for rURN, accessEntries := range resourceAccess {
+		resource, ok := resourcesMap[rURN]
+		if !ok {
+			continue // skip access for resources that not yet added to guardian
+		}
+
+		importedGrants := []*domain.Grant{}
+		accountAccessMap := groupAccessEntriesByAccount(accessEntries)
+		for accountSignature, accessEntries := range accountAccessMap {
+			var accountGrants []*domain.Grant
+			for _, ae := range accessEntries {
+				g := ae.ToGrant(*resource)
+				accountGrants = append(accountGrants, &g)
+			}
+
+			rc := resourceConfigMap[resource.Type]
+			importedGrants = append(importedGrants, reduceGrantsByProviderRole(rc, accountGrants)...)
+
+			for _, g := range importedGrants {
+				if grantsByAccountSignature, ok := activeGrantsMap[rURN]; ok {
+					if grantsByPermissionsKey, ok := grantsByAccountSignature[accountSignature]; ok {
+						g.ID = grantsByPermissionsKey[g.PermissionsKey()].ID
+						delete(grantsByPermissionsKey, g.PermissionsKey())
+					}
+				}
+			}
+		}
+
+		if err := s.repo.BulkUpsert(ctx, importedGrants); err != nil {
+			return nil, fmt.Errorf("inserting imported grants into the db for %q: %w", rURN, err)
+		}
+		newGrants = append(newGrants, importedGrants...)
+	}
+
+	var providerInactiveGrants []*domain.Grant
+	for _, v := range activeGrantsMap {
+		for _, v2 := range v {
+			for _, g := range v2 {
+				g.StatusInProvider = domain.GrantStatusInactive
+				providerInactiveGrants = append(providerInactiveGrants, g)
+			}
+		}
+	}
+	if err := s.repo.BulkUpsert(ctx, providerInactiveGrants); err != nil {
+		return nil, fmt.Errorf("updating grants provider status: %w", err)
+	}
```

Comment:
Can you please put a comment explaining what this is doing?
Why do we need to separate the imported and the inactive one? Can we just make a DB call for all of them?

Reply:
imported grants are fetched from the provider and `inactiveGrants` are the remaining active grants that no longer have access entry in the provider, therefore we marking them as inactive